cjdns-bytes/src/utils.rs

```diff
@@ -1,150 +1,224 @@
-        self.0
-    }
+    impl Writer {
+        pub fn with_capacity(capacity: usize) -> Self {
+            Self(Vec::with_capacity(capacity))
+        }
 
-    pub fn write_u8(&mut self, data: u8) {
-        self.write_slice(&[data]);
-    }
+        pub fn into_vec(self) -> Vec<u8> {
+            self.0
+        }
 
-    pub fn write_u16_be(&mut self, data: u16) {
-        self.write_slice(&data.to_be_bytes());
-    }
+        pub fn write_u8(&mut self, data: u8) {
+            self.write_slice(&[data]);
+        }
 
-    pub fn write_u32_be(&mut self, data: u32) {
-        self.write_slice(&data.to_be_bytes())
-    }
+        pub fn write_u16_be(&mut self, data: u16) {
+            self.write_slice(&data.to_be_bytes());
+        }
 
-    pub fn write_u64_be(&mut self, data: u64) {
-        self.write_slice(&data.to_be_bytes());
-    }
+        pub fn write_u32_be(&mut self, data: u32) {
+            self.write_slice(&data.to_be_bytes())
+        }
 
-    pub fn write_slice(&mut self, data: &[u8]) {
-        self.0.extend_from_slice(data);
-    }
-}
+        pub fn write_u64_be(&mut self, data: u64) {
+            self.write_slice(&data.to_be_bytes());
+        }
 
-#[cfg(test)]
-mod tests {
-    use super::*;
-
-    #[test]
-    fn test_reader() {
-        let bytes = vec![0u8; 64];
-        let mut reader = Reader::new(&bytes);
-
-        let array_32 = reader.read_array_32().expect("bytes are read out");
-        assert_eq!(array_32, [0; 32]);
-        let num64 = reader.read_u64_be().expect("bytes are read out");
-        assert_eq!(num64.to_be_bytes(), [0; 8]);
-        let num32 = reader.read_u32_be().expect("bytes are read out");
-        assert_eq!(num32.to_be_bytes(), [0; 4]);
-        let num16 = reader.read_u16_be().expect("bytes are read out");
-        assert_eq!(num16.to_be_bytes(), [0; 2]);
-        let num8 = reader.read_u8().expect("bytes are read out");
-        assert_eq!(num8, 0);
-        // checking read out of bounds
-        assert!(reader.take_bytes(20).is_err());
-        // reading last bytes
-        reader.take_bytes(17).expect("bytes are read out");
-        assert_eq!(reader.len(), 0);
+        pub fn write_slice(&mut self, data: &[u8]) {
+            self.0.extend_from_slice(data);
+        }
     }
 
-    #[test]
-    fn test_writer() {
-        let initial_capacity = 64;
-        let mut writer = Writer::with_capacity(64);
+    #[cfg(test)]
+    mod tests {
+        use super::*;
 
-        writer.write_u8(0);
-        writer.write_u16_be(0);
-        writer.write_u32_be(0);
-        writer.write_u64_be(0);
-        writer.write_slice(&[0; 49]);
+        #[test]
+        fn test_writer() {
+            let initial_capacity = 64;
+            let mut writer = Writer::with_capacity(64);
 
-        let bytes = writer.into_vec();
-        assert_eq!(bytes.len(), initial_capacity);
-        assert_eq!(bytes, vec![0; initial_capacity]);
-    }
+            writer.write_u8(0);
```

Comment:
Again, don't write zeroes, write unique data, and check it in the end that you get the [0x01, 0x02, 0x03] array as the output.